### PR TITLE
redpanda: generate values.schema.json from go structs

### PR DESCRIPTION
**Pull request by chrisseto**
# Based on PR #1089

This commit introduces a set of go structs that are now used to generate
the values.schema.json files for the redpanda chart.

This is being done to:
- Ensure that values.schema.json is always up to date.
- Allow correct and ergonomic go code to be written in place of helm
  templates.
- Provide a more tenable way of defining, updating, and shaping the
  input values of the redpanda chart.

Efforts were taken to minimize the diff between the original and now
generated jsonschema but there is still a non-trivial diff due to many
invalid entries within the original jsonschema.

Future commits will slowly remove the hacks and shims once the initial
generation hurdle has been cleared.

Rather than reviewing a standard diff. It is recommended to instead view
the diff output by:

`dyff between -i  --exclude-regexp '.*annotations|.*labels' (git show origin/main:charts/redpanda/values.schema.json | psub) ./charts/redpanda/values.schema.json`

- `dyff` will simplify the overall diff.
- `-i` ignores any change in the ordering of keys or elements within an
  array as neither are relevant in the context of a jsonschema.
- `--exclude-regexp '.*annotations|.*labels'` many label and annotation
  fields did not have typed keys. As the Kubernetes API server would
  validate that all values are strings, this is now enforced at the
  schema level.

For reviewers thereof, here is an explanation of the larger diffs:

`properties.storage.properties.tiered.properties.credentialsSecretRef`
was missing a `properties` key in the handwritten schema.

`properties.statefulset.properties.sideCars.properties.configWatcher.controllers`
was incorrectly nested within `configWatcher` _and_ missing a
`properties` key. It's now been hoisted to the correct location within
sideCars and formatted correctly.

`properties.logging` used the key `parameters` instead of `properties`.

### CODE REVIEW

**Comment by chrisseto** [comment]
Here is the output of the `dyff` command from the commit message:

```
     _        __  __
   _| |_   _ / _|/ _|  between /var/folders/jy/5fvcqh9937d82pv9py94t4jc0000gp/T//.psub.PiK8jKcwZa
 / _' | | | | |_| |_       and ./charts/redpanda/values.schema.json
| (_| | |_| |  _|  _|
 \__,_|\__, |_| |_|   returned 20 differences
        |___/

properties.auditLogging.properties.enabledEventTypes
  + one map entry added:
    items:
      type: string

properties.auditLogging.properties.excludedPrincipals
  + one map entry added:
    items:
      type: string

properties.auditLogging.properties.excludedTopics
  + one map entry added:
    items:
      type: string

properties.auth.properties.sasl.properties.users
  - one map entry removed:
    minItems: 0

properties.auth.properties.sasl.properties.users.items
  + one map entry added:
    type: object

properties.commonLabels
  + one map entry added:
    additionalProperties:
      type: string

properties.external.properties.addresses
  + one map entry added:
    items:
      type: string

properties.external.properties.sourceRanges
  + one map entry added:
    items:
      type: string

properties.logging
  - one map entry removed:
    parameters:
      logLevel:
        type: string
        pattern: ^(error|warn|info|debug|trace)$
      usageStats:
        type: object
        required:
        - enabled
        properties:
          enabled:
            type: boolean
  
  + one map entry added:
    properties:
      logLevel:
        type: string
        pattern: ^(error|warn|info|debug|trace)$
      usageStats:
        type: object
        required:
        - enabled
        properties:
          enabled:
            type: boolean
  

properties.nodeSelector
  + one map entry added:
    additionalProperties:
      type: string

properties.statefulset.properties.additionalRedpandaCmdFlags
  + one map entry added:
    items:
      type: string

properties.statefulset.properties.nodeSelector
  + one map entry added:
    additionalProperties:
      type: string

properties.statefulset.properties.sideCars.properties
  + one map entry added:
    controllers:
      type: object
      properties:
        resources: true
        securityContext: true
        enabled:
          type: boolean
        image:
          type: object
          required:
          - tag
          - repository
          properties:
            repository:
              type: string
              default: docker.redpanda.com/redpandadata/redpanda-operator
              pattern: ^[a-z0-9-_/.]+$
            tag:
              type: string
              default: Chart.appVersion
              pattern: "^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$|^$"
    
  

properties.statefulset.properties.sideCars.properties.configWatcher
  - one map entry removed:
    controllers:
      type: object
      properties:
        resources:
          type: object
        enabled:
          type: boolean
        healthProbeAddress:
          type: string
        metricsAddress:
          type: string
        run:
          type: array
        securityContext:
          type: object
        image:
          type: object
          description: "Values used to define the container image to be used for Redpanda"
          required:
          - repository
          - tag
          properties:
            repository:
              type: string
              default: docker.redpanda.com/redpandadata/redpanda-operator
              description: "container image repository"
              pattern: ^[a-z0-9-_/.]+$
            tag:
              type: string
              default: Chart.appVersion
              description: "The container image tag. Use the Redpanda release version. Must be a valid semver prefixed with a 'v'."
              pattern: "^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$|^$"
    
  

properties.statefulset.properties.tolerations
  + one map entry added:
    items: true

properties.statefulset.properties.topologySpreadConstraints.items
  + one map entry added:
    type: object

properties.storage.properties.tiered.properties.credentialsSecretRef.properties.accessKey
  - three map entries removed:     + one map entry added:
    name:                            properties:
      type: string                     name:
    configurationKey:                    type: string
      type: string                     configurationKey:
    key:                                 type: string
      type: string                     key:
                                         type: string

properties.storage.properties.tiered.properties.credentialsSecretRef.properties.secretKey
  - three map entries removed:     + one map entry added:
    name:                            properties:
      type: string                     name:
    configurationKey:                    type: string
      type: string                     configurationKey:
    key:                                 type: string
      type: string                     key:
                                         type: string

properties.tls.properties.certs.patternProperties.^[A-Za-z_][A-Za-z0-9_]*$.properties.issuerRef
  - one map entry removed:
    additionalProperties: false

properties.tolerations
  + one map entry added:
    items:
      type: object


```

File: Taskfile.yaml
Comment (alejandroEsc) [inline Taskfile.yaml:49]: going the reverse of what we had, interesting.

File: charts/redpanda/values_util.go
Comment (alejandroEsc) [inline charts/redpanda/values_util.go:33]: do you mean Enable?

File: charts/redpanda/values_util.go
Comment (alejandroEsc) [inline charts/redpanda/values_util.go:37]: same here?

**Review by alejandroEsc — APPROVED** [review]
A lot to review, two questions, but not blocking.

File: Taskfile.yaml
Comment (chrisseto) [reply Taskfile.yaml:49]: How do you mean? Were we generating something from the json schema previously?

Comment (alejandroEsc) [reply Taskfile.yaml:49]: exactly. I took an existing tool to take schemas from us and create apis.

Comment (chrisseto) [reply Taskfile.yaml:49]: What was the output of that and was it ever automated anywhere? The `values.schema.json` file as it currently exists is missing a good chunk of fields and it's fairly malformed as well.

File: charts/redpanda/values_util.go
Comment (chrisseto) [reply charts/redpanda/values_util.go:33]: No, `Enabled` is the field within it. This is a generic type for anything that's shaped like `{enabled: true}` which I'm calling an "enablable". I'll update the comments before merging!

File: Taskfile.yaml
Comment (alejandroEsc) [reply Taskfile.yaml:49]: No, we never got around to make it automated, ill send you the link so you can see the repo.

Comment (alejandroEsc) [reply Taskfile.yaml:49]: https://github.com/alejandroEsc/generate

File: charts/redpanda/values_util.go
Comment (chrisseto) [reply charts/redpanda/values_util.go]: Eh, I decided to just replaced these with inline structs for now. It's more clear that way. We can always swap back or start using embedding if desired.

**Comment by chrisseto** [comment]
Looks like this is going to be blocked until https://github.com/redpanda-data/helm-charts/pull/1100 get's merged :/

**Review by RafalKorepta — APPROVED** [review]

**Pull request merged**
